File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline

K, N = map(int, input().split())
line = [int(input()) for _ in range(K)]

start = 1       # 1부터 시작
end = max(line)     # 제일 긴 길이의 랜선

while start <= end:
    mid = (start + end) // 2
    cnt = 0

    for i in line:
        cnt += i // mid
    
    if cnt >= N:
        start = mid + 1
    else:
        end  = mid - 1

print(end)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
// baekjoon 13913
#include <bits/stdc++.h>
using namespace std;
typedef pair<int, int> pii;
typedef long long ll;
#define endl "\n"
#define MAX 1025

int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1 ,1};
ll N, K;
vector<ll> v;


int main() {
    ios::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);

    cin >> N >> K;
    for(int i = 0; i < N; i++) {
        int a;
        cin >> a;
        v.push_back(a);
    }

    ll start = 1;
    ll end = 1e15;

    while(start + 1 < end) {
        ll mid = (start + end) / 2;

        ll cnt = 0;
        for(int i = 0; i < N; i++) {
            cnt += v[i] / mid;
        }

        if(cnt >= K) start = mid;
        else end = mid;

    }

    cout << start;
}